crates/ruff_python_parser/src/token_source.rs

```diff
@@ -58,6 +58,23 @@ impl<'src> TokenSource<'src> {
         self.lexer.take_value()
     }
 
+    /// Calls the underlying [`re_lex_logical_token`] method on the lexer and updates the token
+    /// vector accordingly.
+    ///
+    /// [`re_lex_logical_token`]: Lexer::re_lex_logical_token
+    pub(crate) fn re_lex_logical_token(&mut self) {
+        if self.lexer.re_lex_logical_token() {
+            let current_start = self.current_range().start();
+            while self
+                .tokens
+                .last()
+                .is_some_and(|last| last.start() >= current_start)
+            {
+                self.tokens.pop();
+            }
+        }
```

Comment:
Is it possible that more than the last token changes? I would have expected that it is just the last token for which the range and kind changes.

Reply:
Yes, it is possible because this vector contains trivia tokens as well. So, for something like:
```py
if call(foo


    def bar():
        pass
```

There are three `NonLogicalNewline` tokens which needs to be removed and replace with a single `Newline` token.

Reply:
Hmm, I wonder if there's a case where testing for `start() >=` won't be sufficiently accurate. But I think the only zero-width tokens that exist are dedent tokens and we should never see them when calling this function